packages/vmakarichev-datagrok_with_cpp/CallWasmWW/wasm/callWasmUtils.js

```diff
@@ -0,0 +1,710 @@
+const NEW_FLOAT_COLUMN_TYPE = 'newFloatColumn';
+const NEW_INT_COLUMN_TYPE = 'newIntColumn';
+const COLUMN = 'column';
+const CALL_RESULT = '_callResult';
+const NUM_OF_ROWS = 'numOfRows';
+const NUM_OF_COLUMNS = 'numOfColumns';
+const REF = 'ref';
+const VALUE = 'value';
+const TABLE_OF_COLUMNS = 'tableFromColumns';
+const OBJECTS = 'objects';
+const INT_TYPE = 'int';
+const DOUBLE_TYPE = 'double';
+const NUMBER = 'number';
+
+// Type-to-heap correspondence.
+// It is required for JS-module generated by Emscripten,
+// and it is used, when passing array data to/from wasm-functions.
+// More info can be found at the following link:
+// https://becominghuman.ai/passing-and-returning-webassembly-array-parameters-a0f572c65d97
+const heapMap = {
+                 'intColumn': "HEAP32",
+                 'floatColumn': "HEAPF32",
+                 'floatColumns': "HEAPF32",
+                 'newFloatColumns': "HEAPF32",
+                 'intColumns': "HEAP32",
+                 'newIntColumns': "HEAP32",
+                 'newFloatColumn': "HEAPF32",
+                 'newIntColumn': "HEAP32"
+                };
+
+// Type signature to typed array map.
+// It is used, when manipulating column(s).
+const typeMap = {
+                 'intColumn': Int32Array,
+                 'floatColumn': Float32Array,
+                 'floatColumns': Float32Array,
+                 'newFloatColumns': Float32Array,
+                 'intColumns': Int32Array,  
+                 'newIntColumns': Int32Array,
+                 'newFloatColumn': Float32Array,
+                 'newIntColumn': Int32Array
+                }; 
+
+// Type-to-shift map.
+// It is used, when passing array to/from wasm-functions.
+// More info can be found at the following link:
+// https://becominghuman.ai/passing-and-returning-webassembly-array-parameters-a0f572c65d97 
+const shiftMap = {'intColumn': 2, 
+                  'floatColumn': 2, 
+                  'floatColumns': 2,  
+                  'newFloatColumns': 2, 
+                  'intColumns': 2,  
+                  'newIntColumns': 2,
+                  'newFloatColumn': 2,
+                  'newIntColumn': 2
+                 };         
+
+// Get input for C++-function.
+// This function takes specification of arguments (argsSpecification) & input data (inputVals)
+// and returns input that will be further used in cpp/wasm-function.
+export function getCppInput(argsSpecification, inputVals)
+{
+    let cppFuncInput = [];
+    let ref;
+
+    // complete an input for cpp
+    let i = 0;
+    for(let key in argsSpecification){
+        let arg = argsSpecification[key]; 
+        let type = arg.type;
+        
+        // skip auxiliry element
+        if(key == CALL_RESULT)             
+            continue;
+
+        // here, we consider each type of input
+        switch(type){           
+
+            // numbers
+            case NUM_TYPE:                
+            case INT_TYPE:
+            case DOUBLE_TYPE:
```

Comment:
i have a suggestion to repalce the switch code with object literals
it's more maintainable and easy-extendable
something like
```
const typeDefinition = {
        'NUM_TYPE': () => {
          arg.data = inputVals[i];
          i++;
        },
        'FLOAT_COLUMN_TYPE': () => {
          console.log('...');
        },
        'default': () => console.log('...')
      };
```
and so on

Reply:
if there are similar functions for different objects (like NUM_TYPE, INT_TYPE, etc.), you can just write a function for it once and then reuse it in this mapping

Reply:
and then you just do something like `typeDefinition[type]`

Reply:
this will be better just because switches are hardly expandable and also always require a break everywhere, which is a bit confusing

Reply:
Thanks! I'll consider this